### ISSUE TITLE
Ionic quickstarter does not work anymore on NodeJS 10 (3.x)

### ISSUE THREAD

**Issue by michaelsauter** [opendevstack/ods-quickstarters#595]
Error:
```
[0;32m    amazon-ebs: + npm install[0m
[0;32m    amazon-ebs: npm WARN deprecated chokidar@2.1.8: Chokidar 2 will break on node v14+. Upgrade to chokidar 3 with 15x less dependencies.[0m
[0;32m    amazon-ebs: npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142[0m
[0;32m    amazon-ebs: npm WARN deprecated fsevents@1.2.13: fsevents 1 will break on node v14+ and could be using insecure binaries. Upgrade to fsevents 2.[0m
[0;32m    amazon-ebs: npm WARN deprecated uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.[0m
[0;32m    amazon-ebs: npm WARN deprecated har-validator@5.1.5: this library is no longer supported[0m
[0;32m    amazon-ebs: npm WARN deprecated urix@0.1.0: Please see https://github.com/lydell/urix#deprecated[0m
[0;32m    amazon-ebs: npm WARN deprecated resolve-url@0.2.1: https://github.com/lydell/resolve-url#deprecated[0m
[0;32m    amazon-ebs:[0m
[0;32m    amazon-ebs: > core-js@3.12.0 postinstall /tmp/workspace/unitt-cd/unitt-cd-run-unitt-fe-ionic-test-unitt-master/node_modules/core-js[0m
[0;32m    amazon-ebs: > node -e "try{require('./postinstall')}catch(e){}"[0m
[0;32m    amazon-ebs:[0m
[0;32m    amazon-ebs:[0m
[0;32m    amazon-ebs: > core-js-pure@3.13.1 postinstall /tmp/workspace/unitt-cd/unitt-cd-run-unitt-fe-ionic-test-unitt-master/node_modules/core-js-pure[0m
[0;32m    amazon-ebs: > node -e "try{require('./postinstall')}catch(e){}"[0m
[0;32m    amazon-ebs:[0m
[0;32m    amazon-ebs:[0m
[0;32m    amazon-ebs: > @angular/cli@12.0.2 postinstall /tmp/workspace/unitt-cd/unitt-cd-run-unitt-fe-ionic-test-unitt-master/node_modules/@angular/cli[0m
[0;32m    amazon-ebs: > node ./bin/postinstall/script.js[0m
[0;32m    amazon-ebs:[0m
[0;32m    amazon-ebs: npm notice created a lockfile as package-lock.json. You should commit this file.[0m
[0;32m    amazon-ebs: npm WARN notsup Unsupported engine for @angular/forms@12.0.2: wanted: {"node":"^12.14.1 || >=14.0.0"} (current: {"node":"10.24.1","npm":"6.14.12"})[0m
[0;32m    amazon-ebs: npm WARN notsup Not compatible with your version of node/npm: @angular/forms@12.0.2[0m
[0;32m    amazon-ebs: npm WARN notsup Unsupported engine for @angular/core@12.0.2: wanted: {"node":"^12.14.1 || >=14.0.0"} (current: {"node":"10.24.1","npm":"6.14.12"})[0m
[0;32m    amazon-ebs: npm WARN notsup Not compatible with your version of node/npm: @angular/core@12.0.2[0m
[0;32m    amazon-ebs: npm WARN notsup Unsupported engine for @angular/common@12.0.2: wanted: {"node":"^12.14.1 || >=14.0.0"} (current: {"node":"10.24.1","npm":"6.14.12"})[0m
[0;32m    amazon-ebs: npm WARN notsup Not compatible with your version of node/npm: @angular/common@12.0.2[0m
[0;32m    amazon-ebs: npm WARN notsup Unsupported engine for @angular/platform-browser-dynamic@12.0.2: wanted: {"node":"^12.14.1 || >=14.0.0"} (current: {"node":"10.24.1","npm":"6.14.12"})[0m
[0;32m    amazon-ebs: npm WARN notsup Not compatible with your version of node/npm: @angular/platform-browser-dynamic@12.0.2[0m
[0;32m    amazon-ebs: npm WARN notsup Unsupported engine for @angular/router@12.0.2: wanted: {"node":"^12.14.1 || >=14.0.0"} (current: {"node":"10.24.1","npm":"6.14.12"})[0m
[0;32m    amazon-ebs: npm WARN notsup Not compatible with your version of node/npm: @angular/router@12.0.2[0m
[0;32m    amazon-ebs: npm WARN notsup Unsupported engine for @angular/platform-browser@12.0.2: wanted: {"node":"^12.14.1 || >=14.0.0"} (current: {"node":"10.24.1","npm":"6.14.12"})[0m
[0;32m    amazon-ebs: npm WARN notsup Not compatible with your version of node/npm: @angular/platform-browser@12.0.2[0m
[0;32m    amazon-ebs: npm WARN notsup Unsupported engine for @stencil/core@2.5.2: wanted: {"node":">=12.10.0","npm":">=6.0.0"} (current: {"node":"10.24.1","npm":"6.14.12"})[0m
[0;32m    amazon-ebs: npm WARN notsup Not compatible with your version of node/npm: @stencil/core@2.5.2[0m
[0;32m    amazon-ebs: npm WARN notsup Unsupported engine for @angular-devkit/build-angular@12.0.2: wanted: {"node":"^12.14.1 || ^14.0.0","npm":"^6.11.0 || ^7.5.6","yarn":">= 1.13.0"} (current: {"node":"10.24.1","npm":"6.14.12"})[0m
[0;32m    amazon-ebs: npm WARN notsup Not compatible with your version of node/npm: @angular-devkit/build-angular@12.0.2[0m
[0;32m    amazon-ebs: npm WARN notsup Unsupported engine for @angular/cli@12.0.2: wanted: {"node":"^12.14.1 || ^14.0.0","npm":"^6.11.0 || ^7.5.6","yarn":">= 1.13.0"} (current: {"node":"10.24.1","npm":"6.14.12"})[0m
[0;32m    amazon-ebs: npm WARN notsup Not compatible with your version of node/npm: @angular/cli@12.0.2[0m
[0;32m    amazon-ebs: npm WARN notsup Unsupported engine for @angular/compiler-cli@12.0.2: wanted: {"node":"^12.14.1 || >=14.0.0"} (current: {"node":"10.24.1","npm":"6.14.12"})[0m
[0;32m    amazon-ebs: npm WARN notsup Not compatible with your version of node/npm: @angular/compiler-cli@12.0.2[0m
[0;32m    amazon-ebs: npm WARN notsup Unsupported engine for @angular/language-service@12.0.2: wanted: {"node":"^12.14.1 || >=14.0.0"} (current: {"node":"10.24.1","npm":"6.14.12"})[0m
[0;32m    amazon-ebs: npm WARN notsup Not compatible with your version of node/npm: @angular/language-service@12.0.2[0m
[0;32m    amazon-ebs: npm WARN notsup Unsupported engine for @angular/compiler@12.0.2: wanted: {"node":"^12.14.1 || >=14.0.0"} (current: {"node":"10.24.1","npm":"6.14.12"})[0m
[0;32m    amazon-ebs: npm WARN notsup Not compatible with your version of node/npm: @angular/compiler@12.0.2[0m
[0;32m    amazon-ebs: npm WARN notsup Unsupported engine for @angular-devkit/architect@0.1200.2: wanted: {"node":"^12.14.1 || ^14.0.0","npm":"^6.11.0 || ^7.5.6","yarn":">= 1.13.0"} (current: {"node":"10.24.1","npm":"6.14.12"})[0m
[0;32m    amazon-ebs: npm WARN notsup Not compatible with your version of node/npm: @angular-devkit/architect@0.1200.2[0m
[0;32m    amazon-ebs: npm WARN notsup Unsupported engine for @angular-devkit/core@12.0.2: wanted: {"node":"^12.14.1 || ^14.0.0","npm":"^6.11.0 || ^7.5.6","yarn":">= 1.13.0"} (current: {"node":"10.24.1","npm":"6.14.12"})[0m
[0;32m    amazon-ebs: npm WARN notsup Not compatible with your version of node/npm: @angular-devkit/core@12.0.2[0m
[0;32m    amazon-ebs: npm WARN notsup Unsupported engine for @angular-devkit/build-optimizer@0.1200.2: wanted: {"node":"^12.14.1 || ^14.0.0","npm":"^6.11.0 || ^7.5.6","yarn":">= 1.13.0"} (current: {"node":"10.24.1","npm":"6.14.12"})[0m
[0;32m    amazon-ebs: npm WARN notsup Not compatible with your version of node/npm: @angular-devkit/build-optimizer@0.1200.2[0m
[0;32m    amazon-ebs: npm WARN notsup Unsupported engine for @angular-devkit/build-webpack@0.1200.2: wanted: {"node":"^12.14.1 || ^14.0.0","npm":"^6.11.0 || ^7.5.6","yarn":">= 1.13.0"} (current: {"node":"10.24.1","npm":"6.14.12"})[0m
[0;32m    amazon-ebs: npm WARN notsup Not compatible with your version of node/npm: @angular-devkit/build-webpack@0.1200.2[0m
[0;32m    amazon-ebs: npm WARN notsup Unsupported engine for @ngtools/webpack@12.0.2: wanted: {"node":"^12.14.1 || ^14.0.0","npm":"^6.11.0 || ^7.5.6","yarn":">= 1.13.0"} (current: {"node":"10.24.1","npm":"6.14.12"})[0m
[0;32m    amazon-ebs: npm WARN notsup Not compatible with your version of node/npm: @ngtools/webpack@12.0.2[0m
[0;32m    amazon-ebs: npm WARN notsup Unsupported engine for css-minimizer-webpack-plugin@3.0.0: wanted: {"node":">= 12.13.0"} (current: {"node":"10.24.1","npm":"6.14.12"})[0m
[0;32m    amazon-ebs: npm WARN notsup Not compatible with your version of node/npm: css-minimizer-webpack-plugin@3.0.0[0m
[0;32m    amazon-ebs: npm WARN notsup Unsupported engine for open@8.0.2: wanted: {"node":">=12"} (current: {"node":"10.24.1","npm":"6.14.12"})[0m
[0;32m    amazon-ebs: npm WARN notsup Not compatible with your version of node/npm: open@8.0.2[0m
[0;32m    amazon-ebs: npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@~2.3.1 (node_modules/chokidar/node_modules/fsevents):[0m
[0;32m    amazon-ebs: npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@2.3.2: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})[0m
[0;32m    amazon-ebs: npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@^1.2.7 (node_modules/webpack-dev-server/node_modules/chokidar/node_modules/fsevents):[0m
[0;32m    amazon-ebs: npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.13: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})[0m
[0;32m    amazon-ebs: npm WARN notsup Unsupported engine for @angular-devkit/schematics@12.0.2: wanted: {"node":"^12.14.1 || ^14.0.0","npm":"^6.11.0 || ^7.5.6","yarn":">= 1.13.0"} (current: {"node":"10.24.1","npm":"6.14.12"})[0m
[0;32m    amazon-ebs: npm WARN notsup Not compatible with your version of node/npm: @angular-devkit/schematics@12.0.2[0m
[0;32m    amazon-ebs: npm WARN notsup Unsupported engine for @schematics/angular@12.0.2: wanted: {"node":"^12.14.1 || ^14.0.0","npm":"^6.11.0 || ^7.5.6","yarn":">= 1.13.0"} (current: {"node":"10.24.1","npm":"6.14.12"})[0m
[0;32m    amazon-ebs: npm WARN notsup Not compatible with your version of node/npm: @schematics/angular@12.0.2[0m
[0;32m    amazon-ebs: npm WARN ajv-keywords@3.5.2 requires a peer of ajv@^6.9.1 but none is installed. You must install peer dependencies yourself.[0m
[0;32m    amazon-ebs:[0m
[0;32m    amazon-ebs: added 1584 packages from 1342 contributors in 110.678s[0m
[0;32m    amazon-ebs:[0m
[0;32m    amazon-ebs: 128 packages are looking for funding[0m
[0;32m    amazon-ebs:   run `npm fund` for details[0m
[0;32m    amazon-ebs:[0m
[0;32m    amazon-ebs: [Pipeline] sh[0m
[0;32m    amazon-ebs: + ionic build --prod[0m
[0;32m    amazon-ebs: > ng run app:build:production[0m
[0;32m    amazon-ebs: Node.js version v10.24.1 detected.[0m
[0;32m    amazon-ebs: The Angular CLI requires a minimum Node.js version of either v12.14 or v14.15.[0m
[0;32m    amazon-ebs:[0m
[0;32m    amazon-ebs: Please update your Node.js version or visit https://nodejs.org/ for additional instructions.[0m
[0;32m    amazon-ebs:[0m
[0;32m    amazon-ebs: [ERROR] An error occurred while running subprocess ng.[0m
[0;32m    amazon-ebs:[0m
[0;32m    amazon-ebs:         ng run app:build:production exited with exit code 3.[0m
[0;32m    amazon-ebs:[0m
[0;32m    amazon-ebs:         Re-running this command with the --verbose flag may provide more information.[0m
[0;32m    amazon-ebs: [Pipeline] }[0m
[0;32m    amazon-ebs: [Pipeline] // withEnv[0m
[0;32m    amazon-ebs: [Pipeline] }[0m
[0;32m    amazon-ebs: [Pipeline] // stage[0m
[0;32m    amazon-ebs: [Pipeline] }[0m
[0;32m    amazon-ebs: [Pipeline] // wrap[0m
[0;32m    amazon-ebs: [Pipeline] stage[0m
[0;32m    amazon-ebs: [Pipeline] { (odsPipeline error)[0m
```

**Comment by michaelsauter** [comment]
Suggestion: Pin Ionic to a version compatible with NodeJS 10.

**Comment by clemensutschig** [comment]
good lord,... why are they overwriting their own versions every other week..

**Comment by clemensutschig** [comment]
@bljubisic can you take a look, and maybe backport your change from master (to pin the CLI and not just the top level version)

**Comment by bljubisic** [comment]
Hm, Ionic Quickstarter for version 3.x is set to 5.14, which should work with node 10, I will check that. Only master is set to the latest Ionic version.

**Comment by clemensutschig** [comment]
But on master you install the cli.. and not the full piece

**Comment by bljubisic** [comment]
On Master it is version of Ionic 6.x which is renamed to Ionic/cli now. That is the difference, and on master it needs Node 12, which is ok, if it is to be used on ODS4, is that right? On ODS 3.x we have Node 10 and Ionic 5.4.16 which is the last version that can use name Ionic, instead of Ionic/cli. I just checked Ionic 5.4.16 with node 10 and it is working as it should.

**Comment by clemensutschig** [comment]
hmmm . very interesting -did you use the quickstarter, or ?

**Comment by clemensutschig** [comment]
ha .. I digged more into the logs ...

the provisioning worked - the build is failing ... 
```
32m    amazon-ebs: + npm i -g @ionic/cli@6.3.0[0m
[0;32m    amazon-ebs: /home/jenkins/.npm-global/bin/ionic -> /home/jenkins/.npm-global/lib/node_modules/@ionic/cli/bin/ionic[0m
[0;32m    amazon-ebs: + @ionic/cli@6.3.0[0m
[0;32m    amazon-ebs: added 218 packages from 156 contributors in 12.777s[0m
[0;32m    amazon-ebs: [Pipeline] sh[0m
```
for some reason (jenkinsfile.template!), it's installing latest ionic I guess ... ? `https://github.com/opendevstack/ods-quickstarters/blob/master/fe-ionic/Jenkinsfile.template#L25` -- :)

so the bug is in the jenkinsfile.template - to pic the 6.x version

**Comment by bljubisic** [comment]
Hi,

Is there a way to see the whole log? I have tried it on plane ubuntu on docker and Node 10 and everything worked. Maybe there was some additional step that I missed.

**Comment by clemensutschig** [comment]
the whole log is on ods-core / readme linked ... when can we expect a fix here?

**Comment by cschweikert** [comment]
Hi, I could reproduce installation of Angular 12 on my local machine.

I did some exploration on release notes and CLI code. Unfortunatelythe I got pretty much confused by the way they organized the switch from the package `ionic` to `@ionic/cli`. The release notes for example have gaps especially on version the version we use here (`ionic@5.4.16`). I could not even find the word Angular 10, 11 or 12 anywhere within the release notes. That would also make it hard to find commits and code explaining why and since when it is using Angular 12.

Hey, it could also be that it tries to get **latest version** instead of any specific. The documentation is claiming that Ionic support everything starting Angular 6 (see also https://ionicframework.com/docs/angular/overview). Even more important it is highly relying on the ``@angular/cli``. I guess with this it might be much easier for Ionic to support any kind of Angular version and preinstall the latest right away.

But how can we solve the issue? I could not find any parameter to define the actual Angular version when calling ``ionic start ...``. I'd propose to **switch the agent to ``jenkins-agent-nodejs12``** instead. This will most likely cause unit tests to not work anymore (missing Chrome in 3.x). For this I'd propose to **merge the ``jenkins-agent-nodejs12`` from master**.

3.x vs. master:
* ``jenkins-agent-nodejs12`` on 3.x: plain agent with nodejs 12
* ``jenkins-agent-nodejs12`` on master: evolution of ``jenkins-agent-nodejs10-angular`` with node 12 support, Chrome, cypress support, etc.

tl;dr

pull ``jenkins-agent-nodejs12`` from master and use this one instead of ``jenkins-agent-nodejs10-angular``

**Comment by cschweikert** [comment]
merging the node 12 agent from master into 3.x will increase that agent's functionallity, but also its size. @michaelsauter Do you see a problem in this?

**Comment by clemensutschig** [comment]
I think this is the bst option we have - so go from my side? @metmajer any concerns?

**Comment by michaelsauter** [comment]
@cschweikert size would not concern me that much given the current situation.

The concern I have is that Node 12 is (as far as I understand) not fully backwards to Node 10. So far, for ODS versions we strive for non-breaking changes within one major version of ODS. Swapping the Node version between ODS 3.1.0 and a potential 3.2.0 is against our current approach. However as I don't see a better alternative I am for the change in this instance.

**Comment by cschweikert** [comment]
@michaelsauter From my point of view moving the Ionic quickstarter to ``jenkins-agent-nodejs12`` is not a backwards compatibility issue since it doesn't affect already created components. They can keep using the "old" ``jenkins-agent-nodejs10-angular``. Also ``jenkins-agent-nodejs10-angular`` will not be touched and ``jenkins-agent-nodejs12`` 3.x would just learn some tricks from the one on master.

**Comment by michaelsauter** [comment]
@cschweikert Oh right since the name differs that works!

**Comment by bljubisic** [comment]
@cschweikert will you create a PR for this? I totally agree with all of this btw.

**Comment by cschweikert** [comment]
@bljubisic I can offer to do this starting 5th of July.

**Comment by cschweikert** [comment]
I created a branch from 3.x with the fix here: https://github.com/opendevstack/ods-quickstarters/tree/fix/switch-ionic-to-nodejs12
I'm 90% sure that it will work like this, but I won't find the time to test it soon.

@bljubisic Can you try running the tests on this?

**Comment by bljubisic** [comment]
I will test this ASAP and let you know how it goes

**Comment by bljubisic** [comment]
Tested localy on docker ubuntu, and everything works as it should. Confirm the same error with node 10 on the same setup

**Comment by bljubisic** [comment]
I have opened the pR for this, hope its ok @cschweikert

**Comment by cschweikert** [comment]
@bljubisic I guess we can close this issue
